File: docs/experiments/17-neural-network-training/neural_network_training.py
#!/usr/bin/env python3
"""Experiment 17 -- Neural-network training: a chain of transposed GEMMs.

Runs the same algorithm as ``neural_network_training.m``, in the same order,
with the same sizes.  See ``README.md`` for the measurements and the analysis.

    python3 neural_network_training.py

WHAT IT MEASURES.  The third sweep's logistic-regression row has ONE weight
matrix.  Real training has a chain, and the backward pass is a chain of
TRANSPOSED products -- ``a1.T @ d2``, ``d2 @ W2.T``, ``X.T @ d1``.  A
two-layer MLP is the smallest kernel that exercises the whole chain.

WHERE NUMPY HAS A STRUCTURAL ADVANTAGE, STATED PLAINLY.  ``X.T`` is a *view*:
it costs nothing and BLAS is then told to read the buffer transposed.  Both
CAS copy.  ``X.T`` in this file is therefore free and the corresponding
``Transpose[nnX]`` in the ``.m`` file is 2 ms a step, recomputed 100 times
because it is loop-invariant but re-evaluated.  That is a capability
difference, not a tuning difference, and it is the largest single item on this
experiment's roadmap.

BIASES ARE A COLUMN OF X, NOT A SEPARATE VECTOR -- what a performance-minded
implementation does in all three languages, and it keeps the three columns
running identical arithmetic.

DETERMINISM.  The timed run uses random data.  The check trains a
deterministic 256 x 17 -> 8 -> 4 network for 20 steps and compares the
cross-entropy loss, which depends on every weight and gradient along the way.
"""

import time

import numpy as np

# ---- shared reporting helpers (identical in every experiment file) --------


def bench(label, fn, reps=3):
    """One untimed warm-up, then the MINIMUM of `reps` timed runs."""
    fn()
    ts = []
    for _ in range(reps):
        t0 = time.perf_counter()
        fn()
        ts.append(time.perf_counter() - t0)
    print("%-52s%s ms" % (label, round(1000.0 * min(ts), 3)))


def check(label, value):
    print("%-52scheck = %s" % (label, value))


def nest(f, x, m):
    for _ in range(m):
        x = f(x)
    return x


def relu(z):
    return np.maximum(z, 0.0)


# ---- the timed network ---------------------------------------------------

NNB = 1024                 # minibatch
NNI = 785                  # inputs, including the constant bias column
NNH = 128                  # hidden units
NNO = 10                   # classes
NNLR = 0.05                # learning rate

nnX = np.concatenate([np.random.uniform(-1, 1, (NNB, NNI - 1)),
                      np.ones((NNB, 1))], 1)
nnY = np.zeros((NNB, NNO))
nnY[np.arange(NNB), (np.arange(1, NNB + 1) % NNO)] = 1.0

# Deterministic initial weights: no RNG, so the two CAS start identically.
nnW1 = 0.05 * np.sin(np.arange(1, NNI + 1)[:, None] + NNI * np.arange(1, NNH + 1)[None, :])
nnW2 = 0.05 * np.cos(np.arange(1, NNH + 1)[:, None] + NNH * np.arange(1, NNO + 1)[None, :])


def sfmax(z):
    """Row-wise softmax, shifted by the row maximum for numerical stability."""
    ex = np.exp(z - z.max(1)[:, None])
    return ex / ex.sum(1)[:, None]


def mlpstep(w):
    """One SGD step: forward, softmax cross-entropy, backward, update."""
    z1 = nnX @ w[0]                        # 1024x785 @ 785x128
    a1 = relu(z1)
    z2 = a1 @ w[1]                         # 1024x128 @ 128x10
    sm = sfmax(z2)
    d2 = (sm - nnY) / NNB                  # dL/dz2 for cross-entropy
    d1 = (d2 @ w[1].T) * (z1 >= 0)         # backprop through ReLU
    g2 = a1.T @ d2
    g1 = nnX.T @ d1                        # nnX.T is a VIEW here, free
    return [w[0] - NNLR * g1, w[1] - NNLR * g2]


def mlptrain(m):
    return nest(mlpstep, [nnW1, nnW2], m)


# ---- the deterministic check network -------------------------------------

CB, CI, CH, CO, CLR = 256, 17, 8, 4, 0.1
cX = np.concatenate([np.sin(0.3 * (np.arange(1, CB + 1)[:, None]
                                   + CI * np.arange(1, CI)[None, :])),
                     np.ones((CB, 1))], 1)
cY = np.zeros((CB, CO))
cY[np.arange(CB), (np.arange(1, CB + 1) % CO)] = 1.0
cW1 = 0.2 * np.sin(np.arange(1, CI + 1)[:, None] + CI * np.arange(1, CH + 1)[None, :])
cW2 = 0.2 * np.cos(np.arange(1, CH + 1)[:, None] + CH * np.arange(1, CO + 1)[None, :])


def cstep(w):
    z1 = cX @ w[0]
    a1 = relu(z1)
    z2 = a1 @ w[1]
    sm = sfmax(z2)
    d2 = (sm - cY) / CB
    d1 = (d2 @ w[1].T) * (z1 >= 0)
    g2 = a1.T @ d2
    g1 = cX.T @ d1
    return [w[0] - CLR * g1, w[1] - CLR * g2]


def closs(m):
    w = nest(cstep, [cW1, cW2], m)
    sm = sfmax(relu(cX @ w[0]) @ w[1])
    return float(-(cY * np.log(sm + 1e-12)).sum() / CB)


# ---- inference -----------------------------------------------------------

NNQ = 8192
nnXq = np.concatenate([np.random.uniform(-1, 1, (NNQ, NNI - 1)),
                       np.ones((NNQ, 1))], 1)


def mlpinfer():
    return sfmax(relu(nnXq @ nnW1) @ nnW2).sum(1)


def main():
    print("Experiment 17 -- neural-network training")
    print("")

    bench("MLP training, 785-128-10, batch 1024, 100 steps",
          lambda: mlptrain(100))
    check("MLP training (loss after 20 deterministic steps)", closs(20))

    bench("MLP inference, 8192 x 785 forward pass", mlpinfer)
    check("MLP inference (loss after 1 step)", closs(1))

    # One training step, split.  The two lines that dominate the .m file --
    # the loop-invariant transpose and the mixed-dtype mask -- are free and
    # nearly free here, which is what makes this the useful control.
    print("")
    print("-- one training step, split (per step) --")
    sz1 = nnX @ nnW1
    sa1 = relu(sz1)
    sd2 = np.random.rand(NNB, NNO)
    sda = sd2 @ nnW2.T
    sdz = sda * (sz1 >= 0)

    bench("nnX @ W1        (forward GEMM)", lambda: nnX @ nnW1)
    bench("maximum(z1, 0)  (ReLU)", lambda: relu(sz1))
    bench("z2.max(1)       (softmax shift)", lambda: (sa1 @ nnW2).max(1))
    bench("nnX.T           (a VIEW, free)", lambda: nnX.T)
    bench("a1.T", lambda: sa1.T)
    bench("a1.T @ d2", lambda: sa1.T @ sd2)
    bench("da1 * (z1 >= 0) (mask)", lambda: sda * (sz1 >= 0))
    bench("nnX.T @ dz1", lambda: nnX.T @ sdz)
    bench("W1 - lr g1      (update)", lambda: nnW1 - NNLR * (nnX.T @ sdz))


if __name__ == "__main__":
    main()

File: docs/experiments/17-neural-network-training/neural_network_training.m
(* ==========================================================================
   Experiment 17 -- Neural-network training: a chain of transposed GEMMs
   ==========================================================================

   Runs unmodified in Mathilda and in Mathematica 14. See README.md.

       Mathilda      ./Mathilda -file neural_network_training.m
       Mathematica   wolframscript -file neural_network_training.m
       Python        python3 neural_network_training.py

   WHAT IT MEASURES.  The third sweep's logistic-regression row has ONE weight
   matrix.  Real training has a chain, and the backward pass is a chain of
   TRANSPOSED products:

       dW2 = a1^T . d2      dh = d2 . W2^T      dW1 = X^T . d1

   Every one of those is a Transpose feeding a Dot, which is the operation the
   fourth sweep named as its largest remaining application gap.  A two-layer
   MLP is the smallest kernel that exercises the whole chain.

   BIASES ARE A COLUMN OF X, NOT A SEPARATE VECTOR.  That is what a
   performance-minded implementation does in all three languages, and it keeps
   the three columns identical: Wolfram's Plus threads the OUTER level, so
   `matrix + rowVector` is not a NumPy-style broadcast and the honest spelling
   would otherwise have to differ per system.

   DETERMINISM.  The timed run uses random data, which no two systems can
   share.  The check trains a DETERMINISTIC 256 x 17 -> 8 -> 4 network for 20
   steps and compares the cross-entropy loss, which depends on every weight
   and every gradient along the way.
   ========================================================================== *)

(* ---- shared reporting helpers (identical in every experiment file) ------- *)

SetAttributes[bench, HoldRest];

bench[label_String, expr_] := Module[{ts},
  expr;
  ts = Table[First[AbsoluteTiming[expr]], {3}];
  Print[StringPadRight[label, 52], ToString[Round[1000. Min[ts], 0.001]], " ms"]
];

check[label_String, value_] :=
  Print[StringPadRight[label, 52], "check = ", value];

relu[zz_] := Ramp[zz];

(* ---- the timed network -------------------------------------------------- *)

nnb  = 1024;      (* minibatch                                              *)
nni  = 785;       (* inputs, including the constant bias column             *)
nnh  = 128;       (* hidden units                                           *)
nno  = 10;        (* classes                                                *)
nnlr = 0.05;      (* learning rate                                          *)

nnX = Join[RandomReal[{-1, 1}, {nnb, nni - 1}], Table[{1.}, {nnb}], 2];
nnY = Table[If[j == Mod[i, nno] + 1, 1., 0.], {i, 1, nnb}, {j, 1, nno}];

(* Deterministic initial weights: no RNG, so the two CAS start identically. *)
nnW1 = Table[0.05 Sin[1. (i + nni j)], {i, 1, nni}, {j, 1, nnh}];
nnW2 = Table[0.05 Cos[1. (i + nnh j)], {i, 1, nnh}, {j, 1, nno}];

(* Row-wise softmax, shifted by the row maximum for numerical stability.

   Both threadings here rely on Wolfram's outer-level rule: `z - Map[Max, z]`
   pairs row i with the scalar max of row i, and `ex/Total[ex, {2}]` pairs row
   i with its own sum.  Neither is a NumPy broadcast; both are exactly what
   the outer-level rule gives. *)
sfmax[z_] := Module[{ex}, ex = Exp[z - Map[Max, z]]; ex/Total[ex, {2}]];

(* One SGD step: forward, softmax cross-entropy, backward, update. *)
mlpstep[w_] := Module[{z1, a1, z2, sm, d2, d1, g1, g2},
  z1 = nnX . w[[1]];                       (* 1024x785 . 785x128            *)
  a1 = relu[z1];
  z2 = a1 . w[[2]];                        (* 1024x128 . 128x10             *)
  sm = sfmax[z2];
  d2 = (sm - nnY)/nnb;                     (* dL/dz2 for cross-entropy      *)
  d1 = (d2 . Transpose[w[[2]]]) UnitStep[z1];   (* backprop through ReLU    *)
  g2 = Transpose[a1] . d2;
  g1 = Transpose[nnX] . d1;                (* Transpose[nnX] is LOOP-       *)
                                           (* INVARIANT and recomputed here *)
  {w[[1]] - nnlr g1, w[[2]] - nnlr g2}];

mlptrain[m_] := Nest[mlpstep, {nnW1, nnW2}, m];

(* ---- the deterministic check network ------------------------------------ *)

cb = 256; ci = 17; ch = 8; co = 4; clr = 0.1;
cX = Join[Table[Sin[0.3 (i + ci j)], {i, 1, cb}, {j, 1, ci - 1}],
          Table[{1.}, {cb}], 2];
cY = Table[If[j == Mod[i, co] + 1, 1., 0.], {i, 1, cb}, {j, 1, co}];
cW1 = Table[0.2 Sin[1. (i + ci j)], {i, 1, ci}, {j, 1, ch}];
cW2 = Table[0.2 Cos[1. (i + ch j)], {i, 1, ch}, {j, 1, co}];

cstep[w_] := Module[{z1, a1, z2, sm, d2, d1, g1, g2},
  z1 = cX . w[[1]]; a1 = relu[z1]; z2 = a1 . w[[2]];
  sm = sfmax[z2]; d2 = (sm - cY)/cb;
  d1 = (d2 . Transpose[w[[2]]]) UnitStep[z1];
  g2 = Transpose[a1] . d2; g1 = Transpose[cX] . d1;
  {w[[1]] - clr g1, w[[2]] - clr g2}];

closs[m_] := Module[{w, sm},
  w  = Nest[cstep, {cW1, cW2}, m];
  sm = sfmax[relu[cX . w[[1]]] . w[[2]]];
  -Total[cY Log[sm + 1.0*^-12], 2]/cb];

(* ---- inference ---------------------------------------------------------- *)

nnq  = 8192;
nnXq = Join[RandomReal[{-1, 1}, {nnq, nni - 1}], Table[{1.}, {nnq}], 2];
mlpinfer[] := Total[sfmax[relu[nnXq . nnW1] . nnW2], 2];

(* ---- run ---------------------------------------------------------------- *)

Print["Experiment 17 -- neural-network training"];
Print[""];

bench["MLP training, 785-128-10, batch 1024, 100 steps", mlptrain[100]];
check["MLP training (loss after 20 deterministic steps)", closs[20]];

bench["MLP inference, 8192 x 785 forward pass",          mlpinfer[]];
check["MLP inference (loss after 1 step)", closs[1]];

(* ---- one training step, split ------------------------------------------- *)
(*
   Two lines dominate, and neither is arithmetic: Transpose[nnX] is
   loop-invariant and recomputed every step, and the ReLU-derivative mask is a
   mixed float64 x int64 product.  Splitting the step is what shows that;
   see README.md for the roadmap it produces.
*)
Print[""];
Print["-- one training step, split (per step) --"];

sz1 = nnX . nnW1;
sa1 = relu[sz1];
sd2 = RandomReal[{0, 1}, {nnb, nno}];
sda = sd2 . Transpose[nnW2];
sdz = sda UnitStep[sz1];

bench["nnX . W1        (forward GEMM)",  nnX . nnW1];
bench["Ramp[z1]        (ReLU)",          relu[sz1]];
bench["Map[Max, z2]    (softmax shift)", Map[Max, sa1 . nnW2]];
bench["Transpose[nnX]  (LOOP-INVARIANT)", Transpose[nnX]];
bench["Transpose[a1]",                    Transpose[sa1]];
bench["Transpose[a1] . d2",               Transpose[sa1] . sd2];
bench["da1 UnitStep[z1] (mixed f64 x i64)", sda UnitStep[sz1]];
bench["Transpose[nnX] . dz1",             Transpose[nnX] . sdz];
bench["W1 - lr g1      (update)",         nnW1 - nnlr Transpose[nnX] . sdz];
